onnxscript/_framework_apis/torch_2_5.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Stable APIs for PyTorch 2.5."""
+
+from __future__ import annotations
+
+__all__ = [
+    "check_model",
+    "convert_version",
+    "get_torchlib_ops",
+    "optimize",
+    "save_model_with_external_data",
+    "OnnxFunctionMeta",
+]
+
+import dataclasses
+import os
+import pathlib
+from typing import Callable
+
+import onnx
+
+from onnxscript import ir
+from onnxscript.ir import _external_data
+
+# Internal flag. Will go away.
+_TORCH_ONNX_SAVE_EXTERNAL_DATA_WITH_IR = (
+    os.getenv("TORCH_ONNX_OFFLOAD_EXTERNAL_DATA_WITH_IR") == "1"
+)
+
+
+@dataclasses.dataclass(frozen=True)
+class OnnxFunctionMeta:
+    """A wrapper of onnx-script function with additional metadata.
+
+    qualified_name: The qualified name of the aten operator.
+    function: The onnx-script function.
+    domain: The domain of the function.
+    name: The name of the function.
+    is_complex: Whether the function is a complex function.
+    """
+
+    qualified_name: str
+    function: Callable
+    domain: str
+    name: str
+    is_complex: bool = False
+
+
+def optimize(model: ir.Model) -> ir.Model:
+    """Optimize the model."""
+
+    # TODO(justinchuby): Use the optimizer
+    return model
+
+
+def convert_version(model: ir.Model, target_version: int) -> ir.Model:
+    """Convert the model to the specified ONNX opset version."""
+    proto = ir.serde.serialize_model(model)
+    proto = onnx.version_converter.convert_version(proto, target_version)
+    return ir.serde.deserialize_model(proto)
+
+
+def check_model(model: ir.Model) -> None:
+    """Check the model."""
+
+    del model  # Unused yet
+
+
+def save_model_with_external_data(model: ir.Model, model_path: str | os.PathLike) -> None:
+    """Save the model with external data. The model is unchanged after saving."""
```

Comment:
What about tensors stored as attribute values (eg., in Constant ops)? Would be helpful to make them external too. As well as the size-based filters to store only tensors larger than small constants/shapes externally?

Reply:
In practice the constant ops do not carry large buffers or weights (like they used to in the old exporter) in PyTorch, so I think it is easier and more robust to just store the initializers. I think if a user wants to customize the behavior they should load the onnx model back using our tools to have finer control. Right now the options are direct - store the weights with the model, or store the weights externally. What do you think?

Reply:
Easier, but not necessarily more robust? IIUC, you are suggesting using "initalizers" as a proxy/approximation for large tensors, and "constant ops" as a proxy/approximation for small tensors ... but that may not always be true? If small initializers (eg., shapes) get externalized, shape-inference may be impacted, because it works without loading external tensors. On the other hand, optimizers/rewriters might produce Constant ops instead of initializers because it is easier? (Also, because we don't have initializers inside functions, and only constant ops, but this may become less of a concern if and when we get to inlining all functions.)

Reply:
I am fine with adjusting the implementation later on, as long as we are fine with the API here. I think we may be better off making all tensors (initializers and node-attributes) of size above a cutoff limit external, leaving the rest internal. But I am fine if the cutoff limit is hard-coded in this logic for now ... only if we want to expose this to users, does it get a bit more complicated.

Reply:
https://github.com/microsoft/onnxscript/issues/1835